x/pki/types/message_add_pki_revocation_distribution_point.go

```diff
@@ -0,0 +1,183 @@
+	return &MsgAddPkiRevocationDistributionPoint{
+		Signer:               signer,
+		Vid:                  vid,
+		Pid:                  pid,
+		IsPAA:                &isPAA,
+		Label:                label,
+		CrlSignerCertificate: crlSignerCertificate,
+		IssuerSubjectKeyID:   issuerSubjectKeyID,
+		DataUrl:              dataUrl,
+		DataFileSize:         dataFileSize,
+		DataDigest:           dataDigest,
+		DataDigestType:       dataDigestType,
+		RevocationType:       revocationType,
+	}
+}
+
+func (msg *MsgAddPkiRevocationDistributionPoint) Route() string {
+	return pkitypes.RouterKey
+}
+
+func (msg *MsgAddPkiRevocationDistributionPoint) Type() string {
+	return TypeMsgAddPkiRevocationDistributionPoint
+}
+
+func (msg *MsgAddPkiRevocationDistributionPoint) GetSigners() []sdk.AccAddress {
+	signer, err := sdk.AccAddressFromBech32(msg.Signer)
+	if err != nil {
+		panic(err)
+	}
+	return []sdk.AccAddress{signer}
+}
+
+func (msg *MsgAddPkiRevocationDistributionPoint) GetSignBytes() []byte {
+	bz := ModuleCdc.MustMarshalJSON(msg)
+	return sdk.MustSortJSON(bz)
+}
+
+func (msg *MsgAddPkiRevocationDistributionPoint) ValidateBasic() error {
+	_, err := sdk.AccAddressFromBech32(msg.Signer)
+	if err != nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid signer address (%s)", err)
+	}
+
+	err = validator.Validate(msg)
+	if err != nil {
+		return err
+	}
+
+	isDataDigestInTypes := true
+	if msg.DataDigestType != 0 {
+		isDataDigestInTypes = false
+		for _, digestType := range allowedDataDigestTypes {
+			if digestType == msg.DataDigestType {
+				isDataDigestInTypes = true
+
+				break
+			}
+		}
+	}
+
+	if !isDataDigestInTypes {
+		return pkitypes.NewErrInvalidDataDigestType(fmt.Sprintf("invalid DataDigestType: %d. Supported types are: %v", msg.DataDigestType, allowedDataDigestTypes))
+	}
+
+	if msg.RevocationType != allowedRevocationType {
+		return pkitypes.NewErrInvalidRevocationType(fmt.Sprintf("invalid RevocationType: %d. Supported types are: %d", msg.RevocationType, allowedRevocationType))
+	}
+
+	cert, err := x509.DecodeX509Certificate(msg.CrlSignerCertificate)
+	if err != nil {
+		return pkitypes.NewErrInvalidCertificate(err)
+	}
+
+	subjectAsMap := x509.SubjectAsTextToMap(cert.SubjectAsText)
+
+	if *msg.IsPAA {
+		if msg.Pid != 0 {
+			return pkitypes.NewErrNotEmptyPid("Product ID (pid) must be empty for root certificates when isPAA is true")
+		}
+
+		if !cert.IsSelfSigned() {
+			return pkitypes.NewErrRootCertificateIsNotSelfSigned(fmt.Sprintf("CRL Signer Certificate must be self-signed if isPAA is True"))
+		}
+
+		strVid, found := subjectAsMap["Mvid"]
+		if found {
+			vid, err := strconv.ParseInt(strings.Trim(strVid, "0x"), 16, 32)
```

Comment:
It seems this logic is the same for root and non-root cases. Please move to a function and call it.